core/src/main/java/com/linecorp/armeria/client/RetryingRpcClient.java

```diff
@@ -0,0 +1,115 @@
+    private final RetryRequestStrategy<? super RpcRequest, Object> retryStrategy;
+
+    /**
+     * Creates a new {@link Client} decorator that handles failures of an invocation and retries RPC requests.
+     */
+    public static Function<Client<? super RpcRequest, ? extends RpcResponse>, RetryingRpcClient>
+    newDecorator(int retries) {
+        return delegate -> new RetryingRpcClient(delegate,
+                                                 RetryRequestStrategy.alwaysFalse(),
+                                                 retries,
+                                                 Backoff::withoutDelay);
+    }
+
+    /**
+     * Creates a new {@link Client} decorator that handles failures of an invocation and retries RPC requests.
+     */
+    public static Function<Client<? super RpcRequest, ? extends RpcResponse>, RetryingRpcClient>
+    newDecorator(int retries,
+                 RetryRequestStrategy<? super RpcRequest, Object> retryRequestStrategy) {
+        return delegate -> new RetryingRpcClient(delegate, retryRequestStrategy, retries,
+                                                 Backoff::withoutDelay);
+    }
+
+    /**
+     * Creates a new {@link Client} decorator that handles failures of an invocation and retries RPC requests.
+     */
+    public static Function<Client<? super RpcRequest, ? extends RpcResponse>, RetryingRpcClient>
+    newDecorator(int retries,
+                 RetryRequestStrategy<? super RpcRequest, Object> retryRequestStrategy,
+                 Supplier<? extends Backoff> backoffSupplier) {
+        return delegate -> new RetryingRpcClient(delegate, retryRequestStrategy, retries, backoffSupplier);
+    }
+
+    /**
+     * Creates a new instance that decorates the specified {@link Client}.
+     */
+    private RetryingRpcClient(Client<? super RpcRequest, ? extends RpcResponse> delegate,
+                              RetryRequestStrategy<? super RpcRequest, Object> retryStrategy,
+                              int retries, Supplier<? extends Backoff> backoffSupplier) {
+        super(delegate, retries, backoffSupplier);
+        this.retryStrategy = requireNonNull(retryStrategy, "retryStrategy");
+    }
+
+    @Override
+    public RpcResponse execute(ClientRequestContext ctx, RpcRequest req) throws Exception {
+        DefaultRpcResponse responseFuture = new DefaultRpcResponse();
+        retry(maxRetries(), newBackoff(), ctx, req, () -> {
+            try {
+                return delegate().execute(ctx, req);
+            } catch (Exception e) {
+                return new DefaultRpcResponse(e);
+            }
+        }, responseFuture);
+        return responseFuture;
+    }
+
+    private void retry(int times, Backoff backoff, ClientRequestContext ctx, RpcRequest req,
+                       Supplier<RpcResponse> action, DefaultRpcResponse responseFuture) {
+        RpcResponse response = action.get();
+        response.thenAccept(result -> {
```

Comment:
A user is not given with a chance to decide to retry or not when an RPC call fails with an exception; the current code that uses `thenAccept` will always retry automatically. The user will want to have more control on that. How about using `handle`?